backend/pkg/k8cache/k8cache.go

```diff
@@ -0,0 +1,329 @@
+		reader, err := gzip.NewReader(bytes.NewReader(bodyBytes))
+		if err != nil {
+			return "", fmt.Errorf("failed to create gzip reader: %w", err)
+		}
+
+		decompressedBody, err := io.ReadAll(reader)
+		if err != nil {
+			return "", fmt.Errorf("failed to decompress body: %w", err)
+		}
+
+		dcmpBody = decompressedBody
+
+		reader.Close()
+	} else {
+		dcmpBody = bodyBytes
+	}
+
+	return string(dcmpBody), nil
+}
+
+// GenerateKey function helps to generate a unique key based on the request from the client
+// The function accepts url( which includes all the information of request ) and contextID which
+// helps to differentiate in multiple contexts.
+func GenerateKey(url *url.URL, contextID string) (string, error) {
+	namespace, err := ExtractNamespace(url.String())
+	if err != nil {
+		return "", err
+	}
+
+	k := CacheKey{Kind: url.Path, Namespace: namespace, Context: contextID}
+
+	key, err := k.SHA()
+	if err != nil {
+		return "", err
+	}
+
+	return key, nil
+}
+
+// UnmarshalCachedData deserialize a JSON string received from cache
+// back into a CacheResposeData struct. This function is used to recontructing
+// the full HTTP response (status, headers, body) when serving the k8's to the client.
+// this is the essential part as it gives the clarity about the incoming k8;s requests.
+func UnmarshalCachedata(cacheResource string,
+	cachedData CachedResponseData,
+) (CachedResponseData, error) {
+	err := json.Unmarshal([]byte(cacheResource), &cachedData)
+	if err != nil {
+		return CachedResponseData{}, err
+	}
+
+	return cachedData, nil
+}
+
+// This function is used when serving response from cache to ensure the client
+// receives correct metadata about the response.
+func SetHeader(cacheData CachedResponseData, w http.ResponseWriter) {
+	for idx, header := range cacheData.Headers {
+		w.Header()[idx] = header
+	}
+
+	w.WriteHeader(cacheData.StatusCode)
+}
+
+// MarshallToStore serialize a cacheResponseData struct into JSON []byte.
+// This function is used before storing the K8's response data into cache.
+// ensuring a consistent and structured format for all cached entries.
+func MarshalToStore(cacheData CachedResponseData) ([]byte, error) {
+	jsonByte, err := json.Marshal(cacheData)
+	if err != nil {
+		return nil, err
+	}
+
+	return jsonByte, nil
+}
+
+const gzipEncoding = "gzip"
+
+// This ensures that the cached headers accurately reflect the state of the
+// decompressed body that is being stored, and prevents client side decompression
+// issues serving from cache.
+func SetHeadersToCache(responseHeaders http.Header, encoding string) http.Header {
+	cacheHeader := make(http.Header)
+
+	for idx, header := range responseHeaders {
+		if strings.EqualFold(idx, "Content-Encoding") && encoding == gzipEncoding {
+			continue
+		}
+
+		cacheHeader[idx] = append(cacheHeader[idx], header...)
+	}
+
+	return cacheHeader
+}
+
+var (
+	clientsetCache = make(map[string]*kubernetes.Clientset)
+	mu             sync.Mutex
+)
```

Comment:
Since we use the token in cache key of `clientsetCache`, the tokens get expired over time and  there is a possibility that the cache is filled with `clientsets` that are expired. We should store the last use time and update the cache to prune unused clientsets in the cache.